app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```diff
@@ -0,0 +1,570 @@
+        return getRecordForValues(new String[]{rawFieldName}, new Object[]{value});
+    }
+
+    @Override
+    public byte[] readBytes(int id) {
+        Cursor cur = getDbOrThrow().query(table, dataColumns,
+                DatabaseHelper.ID_COL + "=?",
+                new String[]{String.valueOf(id)}, null, null, null);
+
+        InputStream is = null;
+        try {
+            cur.moveToFirst();
+            byte[] serializedObj = cur.getBlob(cur.getColumnIndexOrThrow(DatabaseHelper.DATA_COL));
+            if (serializedObj != null) {
+                return serializedObj;
+            } else {
+                String filename = cur.getString(cur.getColumnIndexOrThrow(DatabaseHelper.FILE_COL));
+                byte[] aesKeyBlob = cur.getBlob(cur.getColumnIndexOrThrow(DatabaseHelper.AES_COL));
+                is = getInputStreamFromFile(filename, aesKeyBlob);
+
+                return StreamsUtil.getStreamAsBytes(is);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to read serialized object from file.", e);
+        } finally {
+            if (is != null) {
+                try {
+                    is.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+            if (cur != null) {
+                cur.close();
+            }
+        }
+    }
+
+    @Override
+    public void write(Persistable persistable) {
+        if (persistable.getID() != -1) {
+            update(persistable.getID(), persistable);
+            return;
+        }
+        SQLiteDatabase db = getDbOrThrow();
+
+        try {
+
+
+            db.beginTransaction();
+
+            long insertedId;
+            ByteArrayOutputStream bos = writeExternalizableToStream(persistable);
+            try {
+                if (blobFitsInDb(bos)) {
+                    // serialized object small enough to fit in db
+                    ContentValues contentValues = helper.getNonDataContentValues(persistable);
+                    contentValues.put(DatabaseHelper.DATA_COL, bos.toByteArray());
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.DATA_COL, contentValues);
+                } else {
+                    // store serialized object in file and file pointer in db
+                    String dataFilePath = HybridFileBackedSqlHelpers.newFileForEntry(dbDir).getAbsolutePath();
+                    setFileAsOrphan(db, dataFilePath);
+                    ContentValues contentValues = helper.getNonDataContentValues(persistable);
+                    contentValues.put(DatabaseHelper.FILE_COL, dataFilePath);
+                    byte[] key = generateKeyAndAdd(contentValues);
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.FILE_COL, contentValues);
+
+                    writeStreamToFile(bos, dataFilePath, key);
+                    unsetFileAsOrphan(db, dataFilePath);
```

Comment:
Before I apply this change to other db ops: Does the ordering with `setFileAsOrphan`/`unsetFileAsOrphan` look correct in terms of making `write` atomic?